File: src/python/oqf.py
import os
from unittest import case

class Question:
    def __init__(self, title="", choices=None, answers=None, points=0, comment="", time=0):
        self.Title = title
        self.Choices = choices or []
        self.Answers = answers or []
        self.Points = points
        self.Comment = comment
        self.Time = time
        self.Required = True

def parse(raw):
    questions = []
    q = None

    for line in map(str.strip, raw.splitlines()):
        if not line:
            continue
        match line[0]:
            case ":":
                q = Question(title=line[1:].strip())
                questions.append(q)
            case "=":
                q.Choices.append(line[1:].strip())
                q.Answers.append(True)
            case "!":
                q.Choices.append(line[1:].strip())
                q.Answers.append(False)
            case '"':
                try:
                    q.Time = int(line[1:])
                except ValueError:
                    q.Time = 0
            case '/':
                try:
                    q.Points = int(line[1:])
                except ValueError:
                    q.Points = 0
            case "#":
                q.Comment = line[1:].strip()
            case "?":
                q.Required = False
            case ";":
                q = None
    return questions

def quiz_fe(filename):
    with open(filename, encoding="utf-8") as f:
        return parse(f.read())

for q in os.listdir("."):
    if q.endswith(".oqf"):
        questions = quiz_fe(q)
        for q in questions:
            print("Title:", q.Title)
            print("Choices:", q.Choices)
            print("Answers:", q.Answers)
            print("---")

File: src/go/oqf.go
package main

import (
	"fmt"
	"log"
	"os"
	"strconv"
)

type Question struct {
	Title    string
	Choices  []string
	Answers  []bool
	Points   int
	Comment  string
	Time     int
	Required bool
}

var DefaultQuestion = Question{
	Time:     -1,
	Required: true,
}

func Parse(raw []byte) ([]Question, error) {
	data := make([]Question, 20)
	q := -1
	a := 0
	t := ""
	flag := ""
	for _, chr := range raw {
		log.Printf("flag = [%s], chr = [%s]\n", flag, string(chr))
		if len(flag) == 0 && chr == ':' {
				data = append(data, DefaultQuestion)
				q += 1
				flag = "T"
				continue
		}
		switch flag[0] {
		case 'T':
			if chr != 0x0A {
				data[q].Title += string(chr)
			} else {
				flag = "QD"
			}
		case 'Q':
			switch flag[1] {
			case 'D':
				switch chr {
				case '=':
					data[q].Choices = append(data[q].Choices, "")
					data[q].Answers = append(data[q].Answers, true)
					flag = "QN"
				case '!':
					data[q].Choices = append(data[q].Choices, "")
					data[q].Answers = append(data[q].Answers, false)
					flag = "QN"
				case '#':
					data[q].Comment = ""
					flag = "QK"
				case ';':
					a = 0
					q += 1
					flag = ""
				case '?':
					data[q].Required = false
				case '"':
					flag = "QT"
				case '/':
					flag = "QP"
				}
			case 'N':
				switch chr {
				case 0x0A:
					a += 1
					flag = "QD"
				case ';':
					a = 0
					flag = ""
				default:
					data[q].Choices[a] += string(chr)
				}
			case 'K':
				if chr == 0x0A {
					flag = "QD"
				} else {
					data[0].Comment += string(chr)
				}
			case 'T':
				if chr == 0x0A {
					time, err := strconv.Atoi(t)
					if err != nil {
						time = -1
					}
					data[q].Time = time
					flag = "QD"
					t = ""
				} else {
					t += string(chr)
				}
			case 'P':
				if chr == 0x0A {
					pts, err := strconv.Atoi(t)
					if err != nil {
						pts = -1
					}
					data[q].Points = pts
					flag = "QD"
					t = ""
				} else {
					t += string(chr)
				}
			}
		}
	}
	return data, nil
}

func main() {
	fmt.Println("File")
	file, err := os.ReadFile(os.Args[1])
	if err != nil {
		panic(err)
	}
	fmt.Println("Parse")
	q, _ := Parse(file)
	fmt.Println(q[0])
}
